### DISCUSSION TITLE
Render with Assumed Non Changing Code

### DISCUSSION THREAD

**Discussion by RoyiAvital** [quarto-dev/quarto-cli#3649] [Q&A]
I use Quarto within VS Code.

I render the HTML output (Reveal.JS) using the `Render` button:

![image](https://user-images.githubusercontent.com/7347975/207305302-370f8641-52f6-44d9-8001-e9606019867a.png)

I can see that when I render over and over with short period in between the process is much faster as it assumes the code sections are not changed, hence skip the process of running the code.
Yet, when there is a longer period from my last render, even if no code is changed, it will rerun the code.

I was wondering if there is a command within VS Code or on the command line that can force render mode with assuming the current code hasn't changed form the last time the render has happened.

This will make thing much faster when working with Quarto.

**Comment by mcanouil** [comment]
The Quarto execute.freeze option is what you are looking for.

**Reply by RoyiAvital** [reply]
I tried this, it still reconnect to the kernel.
Could it be working on Linux only? I'm on Windows.

**Reply by mcanouil** [reply]
I do not know exactly when the freeze option intervene, but what matters is not that the kernel is launched but whether or not any chunk are actually computed again.

Anyway, as usual a minimal reproducible example helps to report an issue and/or exhibit a behaviour we'd like to change.

PS: I am a VSCode user and I never encountered or noticed the behaviour you described.

**Reply by RoyiAvital** [reply]
Are you on Windows?
By definition, any time you hit the `Render` button it renders the document. Part of the process it runs the code (Convert to `ipython`). It does that even if the code hasn't changed. Unless the time window between renders is really small.

**Reply by mcanouil** [reply]
Windows and MacOS.

Here you add ipython, meaning you are not talking about any kernel, thus a small reproducible example of what you try to describe would really help to get the whole picture and also help me (and others) to try to reproduce.

**Reply by RoyiAvital** [reply]
@mcanouil , I mean there is code (`Python`) in my `qmd` notebook. Nothing external.
This will do it:

````md
---
title: "Some Title"
author: "Some Author"
editor:
  render-on-save: false
format: 
  revealjs:
    width: 1200
    height: 800
    footer: "Some Footer"
    logo: "https://i.imgur.com/xxxx.png"
    toc: true
    toc-title: Agenda
    toc-depth: 2
    margin: 0.05
    navigation-mode: vertical
revealjs-plugins:
  - attribution
filters:
  - roughnotation
---

```{python}
#| echo: false

# Miscellaneous
import itertools
import os
import random

# Numerical
import numpy as np
import scipy as sp

from scipy.stats import bernoulli, binom
from scipy.stats import beta

from scipy.special import comb

# Data
import pandas as pd

# Visualization
from matplotlib.animation import FuncAnimation
import matplotlib.pyplot as plt
import seaborn as sns

# Jupyter
from IPython.display import HTML

# Configuration
%matplotlib inline


seedNum = 512
np.random.seed(seedNum)
random.seed(seedNum)

sns.set_theme()
```

# Basic Concepts

## Some Concepts

 - AAA.
 - BBB.

## The Course

:::{.incremental}
 * CCC.
 * DDD.
:::

## Some Code

```{python}
#| layout-ncol: 2
%matplotlib inline

numSamples  = 100
paramP      = 0.4

vSamples = np.random.binomial(1, paramP, numSamples)

hF, hA = plt.subplots(figsize = (20, 10))
sns.swarmplot(x = vSamples, ax = hA)
hA.set_xticks([0, 1])
hA.set_title(f'{numSamples} of Bernoulli Random Variable')
hA.set_xlabel(f'Value')

hF, hA = plt.subplots(figsize = (20, 10))
sns.barplot(x = [0, 1], y = [paramP, 1 - paramP], ax = hA)
hA.set_title(f'Bernoulli Distribution - p = {paramP}')
hA.set_xlabel(f'Trial Result')
hA.set_ylabel(f'Probability')

plt.show()
```
````

Do as following:

1. Render it on Windows using the button.
2. Do some changes to text (Not code!).
3. Click render within 5 seconds from (1) again. No code will be rerun.
4. Do some changes to text (Not code!).
5. Wait 3 minutes, click `Render`.

In step (5) it will re initiate the run of the code though nothing has changed.
Namely it will start the kernel and regenerate the output though nothing has changed.

What I'm after is a `Render` button which doesn't rerun the code. Namely it assumes no changes to code.

**Reply by RoyiAvital** [reply]
@mcanouil , I made the code more readable.

**Reply by mcanouil** [reply]
I'll try to reproduce when I'll get some spare time ;)

**Reply by mcanouil** [reply]
I think I can reproduce ...

Contrary to a Jupyter Notebook, here Quarto recreates it at each `quarto render`, thus everything is new in it, not just text.
It's likely to be the reason of what you are seeing.

Note that you could use cache to avoid this as described in the documentation, see <https://quarto.org/docs/projects/code-execution.html#cache>.

**Reply by RoyiAvital** [reply]
For some reason I remember the cache option being only available on Linux or so.  
Anyhow, it would be great to have this feature working to make things easier when working on a document with a lot of code.

**Reply by cscheid** [reply]
> For some reason I remember the cache option being only available on Linux or so.

Just to clarify, that's definitely not the case. If you could point us to where in the documentation you read that, I'd be happy to fix it.

**Reply by RoyiAvital** [reply]
@cscheid , I am not sure about this. The important thing is you clarify the current situation.

**Reply by cscheid** [reply]
It looks like your document isn't using the cache.

You need to enable caching like [the documentation](https://quarto.org/docs/projects/code-execution.html#cache) suggests:

```yaml
execute:
  cache: true
```

There's a difference between:
- our [execution daemon](https://quarto.org/docs/computations/python.html#kernel-daemon) (which speeds up the _starting of the jupyter process_)
- our [freeze option](https://quarto.org/docs/projects/code-execution.html#freeze) (which _prevents the running of any engine, jupyter or otherwise_)
- the _cache_ option, which reuses results from particular cells, using [the knitr or jupyter cache](https://quarto.org/docs/projects/code-execution.html#cache)

In passing, I'd also ask that you make an additional effort to be kind to the developers as we respond here. When you write "The important thing is", your message implies that other things the devs do are not important. Everyone has a lot on their plate.

**Reply by RoyiAvital** [reply]
@cscheid , I meant the important thing is that, you, the knowledgeable confirmed the `cache` works on Windows (It seems I mixed it with the daemon you mentioned).

I am sorry if my choice of words was not clear. 
I have nothing but respect to this project and those who keep pushing it forward.
